### PR TITLE
(chore) O3-3551: Use latest Actions in GH Actions

### PR DESCRIPTION
**Pull request by Bcoderx6**
https://openmrs.atlassian.net/browse/O3-3551

This PR updates several GitHub Actions to their latest versions for improved performance, security, and compatibility.

Changes:
- Updated `actions/checkout` from `v4` to `v4.1.7`
- Updated `actions/setup-java` from `v4` to `v4.2.1`
- Updated `actions/cache` from `v4` to `v4.0.2`
- Updated `actions/upload-artifact` from `v4` to `v4.3.4`

These updates ensure we are using the most recent enhancements and fixes provided by the maintainers.

### CODE REVIEW

**Comment by Bcoderx6** [comment]
@jayasanka-sack !

File: .github/workflows/build-2.x.yaml [hunk 1]
```diff
@@ -28,14 +28,14 @@ jobs:
           - 8
     runs-on: ${{ matrix.platform }}
     steps:
-      - uses: actions/checkout@v4
+      - uses: actions/checkout@v4.1.7
```
Comment (jayasanka-sack) [inline .github/workflows/build-2.x.yaml:31]:
```suggestion
      - uses: actions/checkout@v4
```
Using `@v4` will automatically use the latest minor and patch version within the v4 major version series.

**Review by jayasanka-sack — CHANGES_REQUESTED** [review]
Thanks for the PR @Bcoderx6 ! Please check the comment below.

**Comment by Bcoderx6** [comment]
@jayasanka-sack ! can u check now i fixed it

File: .github/workflows/build-2.x.yaml (hunk 1 above)
Comment (jayasanka-sack) [reply .github/workflows/build-2.x.yaml]: @Bcoderx6 this comment applies to all other places in this file.

**Comment by jayasanka-sack** [comment]
@Bcoderx6 check this comment: https://github.com/openmrs/openmrs-core/pull/4690#discussion_r1680776206

**Comment by Bcoderx6** [comment]
@jayasanka-sack !

**Review by jayasanka-sack — APPROVED** [review]
Thanks @Bcoderx6!

**Pull request merged**